xptifw/src/xpti_trace_framework.cpp

```diff
@@ -976,11 +977,34 @@ class Framework {
   }
 
   static Framework &instance() {
-    static Framework *framework = new Framework();
-    return *framework;
+    Framework *TmpFramework = MInstance.load(std::memory_order_relaxed);
+    std::atomic_thread_fence(std::memory_order_acquire);
+    if (TmpFramework == nullptr) {
+      std::lock_guard<utils::SpinLock> Lock{MSingletoneMutex};
+      TmpFramework = MInstance.load(std::memory_order_relaxed);
+      if (TmpFramework == nullptr) {
+        TmpFramework = new Framework();
+        std::atomic_thread_fence(std::memory_order_release);
+        MInstance.store(TmpFramework, std::memory_order_relaxed);
+      }
+    }
+
+    return *TmpFramework;
```

Comment:
I assume this change attempts to to make the singleton thread-safe? Is this based on some known pattern? If such, please name it in the comment. Maybe we can make an utility out of it? Alternatively, can't we just use "std::call_once"?

Reply:
This is called double-checked lock. The reason I use it is that the framework library can be loaded/unloaded multiple times throughout application lifetime. Each time we must be able to initialize the framework from scratch, and that is why `std::call_once` is suitable here.